src/johnsession.cpp

```diff
@@ -0,0 +1,309 @@
+{
+    m_sessionName = sessionName;
+    m_sessionGroup = "Sessions/" + m_sessionName;
+    m_settings = settings;
+}
+
+static QString JohnSession::defaultSessionDir()
+{
+    return QDir::home().filePath(QLatin1String(".john/sessions/"));
+}
+
+QString JohnSession::recFile()
+{
+    if (sessionName.isEmpty())
+        return "";
+    else
+        return defaultSessionDir() + m_sessionName;
+}
+
+JohnSession::~JohnSession()
+{
+
+}
+
+bool JohnSession::load()
+{
+    bool success = m_settings.contains(m_sessionGroup);
+    if (!success)
+        return false;
+
+    m_settings.beginGroup(m_sessionGroup);
+    m_format = m_settings.value("formatJohn").toString();
+    QString mode = m_settings.value("mode").toString();
+    if (mode == "single") {
+        m_mode = SINGLECRACK_MODE;
+        // External mode, filter
+        if (m_settings.contains("singleCrackExternalName")) {
+            m_externalName = m_settings.value("singleCrackExternalName").toString();
+        }
+    } else if (mode == "wordlist") {
+        m_mode = WORDLIST_MODE;
+        m_wordlistFile = m_settings.value("wordlistFile").toString();
+        //Rules
+        if (m_settings.value("isUsingWordListRules").toBool() == true) {
+            m_rules = "";
+        } else if (m_settings.contains("rules")) {
+            m_rules = m_settings.value("rules");
+        }
+        // External mode, filter
+        if (m_settings.contains("worldListExternalName")) {
+            m_externalName = m_settings.value("worldListExternalName").toString();
+        }
+    } else if (mode == "incremental") {
+        m_mode = INCREMENTAL_MODE;
+        // "Incremental" mode
+        // It could be with or without name.
+        if (m_settings.contains("incrementalModeName")) {
+            m_charset = m_settings.value("incrementalModeName").toString();
+        }
+        // External mode, filter
+        if (m_settings.contains("incrementalExternalName")) {
+            m_externalName = m_settings.value("incrementalExternalName").toString();
+        }
+    } else if (mode == "external") {
+        m_mode = EXTERNAL_MODE;
+        m_externalName = m_settings.value("externalModeName").toString();
+    } else {
+        m_mode = DEFAULT_MODE;
+    }
+
+    // Selectors
+    if (m_settings.contains("limitUsers")) {
+        m_limitUsers = m_settings.value("limitUsers").toString();
+    }
+    if (m_settings.contains("limitGroups")) {
+        m_limitGroups = m_settings.value("limitGroups").toString();
+    }
+    if (m_settings.contains("limitShells")) {
+        m_limitShells = m_settings.value("limitShells").toString();
+    }
+    if (m_settings.contains("limitSalts")) {
+        m_limitSalts = m_settings.value("limitSalts").toInt();
+    }
+    // Advanced options
+    if (m_settings.contains("nbForkProcess")) {
+        m_nbProcess = m_settings.value("nbForkProcess").toInt();
+    }
+    m_ui->spinBox_nbOfOpenMPThread->setValue(m_settings.value("OMP_NUM_THREADS").toInt());
```

Comment:
Where's is this m_ui coming from? JohnSession shouldn't know about it. If it's work in progress ignore.